### PR TITLE
endpoint: add recursion limit check

### PR DESCRIPTION
**Pull request by BusyJay**
This is a rewrite of pr #2424. Instead of checking it in executors, check it at the very beginning of deserialization.

Because pr stepancheg/rust-protobuf#248 is not merged (and released) yet, so use a branch for now.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -97,6 +97,9 @@ optional = true
 [dependencies.fail]
 git = "https://github.com/pingcap/fail-rs.git"
 
+[replace]
+"protobuf:1.4.1" = { git = "https://github.com/busyjay/rust-protobuf.git", branch = "set-recurse-limit" }
```
Comment (siddontang) [inline Cargo.toml]: do we need to update kvproto?

Comment (BusyJay) [reply Cargo.toml]: cargo will handle this automatically for the configuration in replace section.

File: src/coprocessor/endpoint.rs [hunk 2]
```diff
@@ -821,4 +838,34 @@ mod tests {
...
+        let mut dag = DAGRequest::new();
+        dag.mut_executors().push(e);
+        let mut req = Request::new();
+        req.set_tp(REQ_TYPE_DAG);
+        req.set_data(dag.write_to_bytes().unwrap());
+        RequestTask::new(req.clone(), box move |_| unreachable!(), 100);
+        RequestTask::new(
+            req,
+            box move |res| {
+                let s = format!("{:?}", res);
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: better to check error and then the error msg.

File: src/coprocessor/endpoint.rs [hunk 3]
```diff
@@ -821,4 +838,34 @@ mod tests {
...
         panic!("suppose to get ServerIsBusy error.");
     }
+
+    #[test]
+    fn test_stack_guard() {
+        let mut expr = Expr::new();
+        for _ in 0..10 {
+            let mut e = Expr::new();
+            e.mut_children().push(expr);
+            expr = e;
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: can height 10 cause stack > 100?

File: src/coprocessor/endpoint.rs (hunk 2 above)
Comment (BusyJay) [reply src/coprocessor/endpoint.rs]: The type of res is a protobuf message, no way to check the exact type of the error.

File: src/coprocessor/endpoint.rs (hunk 3 above)
Comment (BusyJay) [reply src/coprocessor/endpoint.rs]: Maybe and may not. The check here is not exact the level of stack but just the nested level of messages.

**Comment by siddontang** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @AndreMouche

**Comment by siddontang** [comment]
PTAL @hicqu

**Comment by overvenus** [comment]
Why not update `etc/config-template.toml`?

**Comment by siddontang** [comment]
CI failed @BusyJay

**Pull request merged**